src/uuid/builder/uuidbuilder.rs

```diff
@@ -183,52 +114,29 @@ mod tests {
     use super::*;
 
     #[test]
-    fn test_string_to_obj_conversions() {
-        let str1 = "01868381-1590-8000-cfe2-68135f43b363";
-        let uuid = Uuid::parse_str(str1).unwrap();
-        let str2 = uuid.to_string();
-        assert_eq!(str1, str2);
+    fn test_build_uuid8_with_instant() {
+            let instant = 0x18C684468F8u64; // Thu, 14 Dec 2023 12:19:23 GMT
+            let uprotocol_uuid = UUIDv8Builder::new().build_with_instant(instant);
+        
+        // instant, version (8) and counter (000) should show up in UUID
+        assert!(uprotocol_uuid.to_hyphenated_string().starts_with("018c6844-68f8-8000-"));
+        // variant is 0b10
+        assert_eq!(uprotocol_uuid.lsb >> 62, 0b10);
     }
 
     #[test]
     fn test_obj_to_string_conversions() {
-        let uuid_factory = UUIDv8Builder::new();
-        let uuid1 = uuid_factory.build();
-        let str1 = uuid1.to_string();
-        let uuid2 = Uuid::parse_str(&str1).unwrap();
-        assert_eq!(str1, uuid2.to_string());
+        let uuid1 = UUIDv8Builder::new().build();
+        let str1 = uuid1.to_hyphenated_string();
+        let uuid2 = uuid::Uuid::parse_str(&str1).unwrap();
+        assert_eq!(str1, uuid2.as_hyphenated().to_string());
     }
 
     #[test]
     fn test_uuid_for_constant_random() {
         let factory = UUIDv8Builder::new();
         let uuid1 = factory.build();
         let uuid2 = factory.build();
-        // assert_eq!(uuid1.to_fields_le().3, uuid2.to_fields_le().3); // Check that the "node" field (least significant 64 bits) is the same
         assert_eq!(uuid1.lsb, uuid2.lsb);
     }
```

Comment:
I enjoyed reading over how Rust atomics are used here within the `UUIDv8Builder` to accomplish keeping a globally consistent ordering of how the UUIDs would be generated.

I wrote a unit test here to attempt to force as much contention as possible with 15 threads spun up and then waiting based on a `Barrier` till all of them are immediately before the call to `build_with_instant`.

Was cool to see it working

```rust
    use std::sync::{Arc, Barrier};
    use std::thread;

    #[test]
    fn test_uuid_atomics() {
        let thread_count_for_contention_test = 15;
        let instant = 0x18C684468F8u64; // Thu, 14 Dec 2023 12:19:23 GMT
        let builder = Arc::new(UUIDv8Builder::new());
        let barrier = Arc::new(Barrier::new(thread_count_for_contention_test));
        let mut handles = vec![];

        for _ in 0..thread_count_for_contention_test {
            let barrier_clone = barrier.clone();
            let builder_clone = Arc::clone(&builder);

            let handle = thread::spawn(move || {
                barrier_clone.wait();
                let _ = builder_clone.build_with_instant(instant);
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.join().unwrap();
        }

        let uprotocol_uuid_last_one = builder.build_with_instant(instant);
        assert!(uprotocol_uuid_last_one.to_hyphenated_string().starts_with("018c6844-68f8-800f"));
    }
```

Reply:
while this looks interesting from a code perspective, I wonder if it actually *proves* the desired behavior or simply produces *anecdotal evidence*, given the *not very deterministic* scheduling of the threads ...

Any thought on that?

Reply:
Yeah I thought about this a bit. Certainly the ordering that the threads are able to get access to the builder is non-deterministic in this test. The test is: "During contention among threads accessing the builder, do the atomics ensure that we have incremented the counter the appropriate number of times?"

I was considering collecting all the threads' UUIDs and confirming that we got all the counter content we were expecting, which I suppose would be a more thorough test, but opted for this "check the last counter" approach.

What do you think?